include/roboptim/core/function.hh

```diff
@@ -47,6 +47,18 @@
 
 # include <roboptim/core/detail/utility.hh>
 
+// We should probably move these templates out of function.hh
+// Flags used for type checking inside the RobOptim function architecture
+# define ROBOPTIM_IS_DIFFERENTIABLE 1UL << 0
+# define ROBOPTIM_IS_TWICE_DIFFERENTIABLE 1UL << 1
+# define ROBOPTIM_IS_LINEAR 1UL << 2
+# define ROBOPTIM_IS_NUMERIC_LINEAR 1UL << 3
+# define ROBOPTIM_IS_QUADRATIC 1UL << 4
+# define ROBOPTIM_IS_NUMERIC_QUADRATIC 1UL << 5
+# define ROBOPTIM_IS_CONSTANT 1UL << 6
+# define ROBOPTIM_IS_POLYNOMIAL 1UL << 7
```

Comment:
Do we need an `unsigned long` right now, which forces us to use defines rather than a (slightly) more readable enum (using `int`) such as:

```cpp
enum FunctionFlag
{
    A = 1 << 0,
    B = 1 << 1, 
    C = 1 << 2,
    D = 1 << 3
};
```

IIRC, in C++11 we can do `enum FunctionFlag: unsigned long {A, B, C, D};` to change the type.